src/renderer/components/panel.ts

```diff
@@ -202,28 +202,43 @@ export default class ComponentPanel {
     this.particles[3].create(this)
   }
 
-  swap_l_execute    () { if (this.counter <= 0) { this.change_state(SWAPPING_L) } }
-  swap_r_execute    () { if (this.counter <= 0) { this.change_state(SWAPPING_R) } }
-  land_execute      () { if (this.counter <= 0) { this.change_state(STATIC) } }
-  swapping_l_execute() { if (this.counter <= 0) { this.state = STATIC /*TODO: use FSM here*/ } }
-  swapping_r_execute() { if (this.counter <= 0) { this.state = STATIC /*TODO: use FSM here*/ } }
-  hang_execute      () { if (this.counter <= 0) { this.change_state(FALL) } }
-
-  swapping_r_enter() {
-    this.counter = TIME_SWAP
-  }
-
-  hang_enter() {
-    this.counter = 10
-  }
-
+  hang_enter()   { this.counter = 10 }
+  hang_execute() { if (this.counter <= 0) { this.change_state(FALL) } }
+  
+  swap_l_execute() { if (this.counter <= 0) { this.change_state(SWAPPING_L) } }
+  swap_r_execute() { if (this.counter <= 0) { this.change_state(SWAPPING_R) } }
+  
   swapping_l_enter() {
     const i1 = this.kind
     const i2 = this.right.kind
     this.kind       = i2
     this.right.kind = i1
     this.counter = TIME_SWAP
   }
+  
+  swapping_r_enter() { this.counter = TIME_SWAP }
+
+  swapping_l_execute() { 
+    if (this.counter <= 0) { 
+        if (this.under === blank ? false : this.under.state === STATIC && this.under.kind === null) {
```

Comment:
since this is repeated as well as on line `234` we can abstract this in a function to make it self documenting such as: `should_hang_with_nothing_under`

Reply:
The same idea is how we have all these edge cases in fuctions as such:

```ts
  /** 
   *  A panel is only considered static stable if it is STATIC and has a kind assigned
   * */
  get static_stable() { return  this.state === STATIC && this.kind !== null }

  /** 
   *  A panel is only considered empty if it is STATIC, SWAP_
   *  and has no kind assigned
   * */
  get empty() {
    return  this.state === STATIC && this.kind === null 
  }

  get hidden_during_garbage_static(){
    return this.state === GARBAGE &&
           this.garbage.state === STATIC
  }

  get empty_when_swapping() {
    return (this.state === SWAP_R     ||
            this.state === SWAP_L     ||
            this.state === SWAPPING_L ||
            this.state === SWAPPING_R) && this.kind === null
  }
  /**
   * A panel can be hidden but not empty this only happens in the case
   * when panels are clearing and are temporarily invisible
   * */
  get hidden_during_clear() { return this.state === CLEAR && this.time_cur >= this.time_pop  }

  /*
   * In order to determine danger is true we need a panel that is considered
   * stable. A stable panel is one that is not falling, landing or clearing, hanging
   * So a static panel is
   *   * swapping with a kind
   *   * static with a kind
   *   * garbage thats static
   */
  get stable(){
    if (this.state === STATIC     && this.kind !== null) { return true }
    if (this.state === SWAP_R     && this.kind !== null) { return true }
    if (this.state === SWAPPING_R && this.kind !== null) { return true }
    if (this.state === SWAPPING_L && this.kind !== null) { return true }
    if (this.state === GARBAGE    && this.garbage.state === STATIC) { return true }
    return false
  }
```